File: src/model/item/ItemType.java
package model.item;

import core.ConstantStore;

public enum ItemType {
	APPLE ("APPLE", true, true, false, false),
	BREAD ("BREAD", true, false, false, false),
	AMMO ("AMMO", false, false, false, false),
	GUN ("GUN", false, false, false, false),
	MEAT ("MEAT", true, false, false, false),
	SONIC ("SONIC", false, false, false, false),
	WAGON ("WAGON", false, false, false, false),
	WHEEL ("WHEEL", false, false, false, true),
	OX ("OX", false, false, true, false),
	TOOLS ("TOOLS", false, false, false, true),
	AXLE ("AXLE", false, false, false, true),
	HORSE ("HORSE", false, false, true, false),
	MULE ("MULE", false, false, true, false),
	STRANGEMEAT ("STRANGE_MEAT", true, false, false, false);
	
	private final String name;
	private final String pluralName;

	private final String description;
	
	private final int cost;
	
	private final double weight;
	
	private final boolean isFood, isPlant, isAnimal, isTool;
	
	private final int factor;
	private final int necessaryQuality;
			
	/**
	 * Makes the item type
	 * @param name Name of the item type
	 * @param description Description of the item type
	 * @param cost Cost of the item type
	 * @param weight Weight of the item type
	 * @param isFood If the item is food
	 * @param isPlant If the item is a plant
	 * @param foodFactor The multiplier for the food (if it is one, else 0)
	 */
	private ItemType (String type, boolean isFood, boolean isPlant, boolean isAnimal, boolean isTool) {
		this.name = ConstantStore.get("ITEMS", type + "_NAME");
		this.pluralName = ConstantStore.get("ITEMS", type + "_PLURAL_NAME");
		
		this.description = ConstantStore.get("ITEMS", type + "_DESCRIPTION");
		this.cost = Integer.parseInt(ConstantStore.get("ITEMS", type + "_COST"));
		this.weight = Double.parseDouble(ConstantStore.get("ITEMS", type + "_WEIGHT"));
		this.isFood = isFood;
		this.isPlant = isPlant;
		this.isAnimal = isAnimal;
		this.isTool = isTool;
		if(isFood) {
			this.factor = Integer.parseInt(ConstantStore.get("ITEMS", type + "_FOOD_FACTOR"));
		} else if (isTool){
			this.factor = Integer.parseInt(ConstantStore.get("ITEMS", type + "_REPAIR_FACTOR"));
		} else {
			this.factor = 0;
		}
		if (ConstantStore.get("ITEMS", type + "_NECESSARY_QUALITY") != null) {
			this.necessaryQuality = Integer.parseInt(ConstantStore.get("ITEMS", type + "_NECESSARY_QUALITY"));
		} else {
			this.necessaryQuality = 0;
		}
	}
	
	/**
	 * The name of the type
	 * @return The name
	 */
	public String getName() {
		return name;
	}
	
	/**
	 * The plural name of the type
	 * @return The plural name
	 */
	public String getPluralName() {
		return pluralName;
	}
	
	/**
	 * The description of the type
	 * @return The description
	 */
	public String getDescription() {
		return description;
	}
	
	/**
	 * The cost of the type
	 * @return The cost
	 */
	public int getCost() {
		return cost;
	}
	
	/**
	 * The weight of the type
	 * @return The weight
	 */
	public double getWeight() {
		return weight;
	}
	
	/**
	 * Whether or not the item is food
	 * @return Is the item food?
	 */
	public boolean isFood() {
		return isFood;
	}
	
	/**
	 * Whether or not the item is a tool
	 * @return Is the item a tool?
	 */
	public boolean isTool() {
		return isTool;
	}
	
	/**
	 * The multiplier for the food
	 * @return The multiplier
	 */
	public int getFactor() {
		return factor;
	}
	
	/**
	 * Is the item a plant
	 * @return Is it a plant?
	 */
	public boolean isPlant() {
		return isPlant;
	}

	/**
	 * Is the item an animal?
	 * @return Is it an animal
	 */
	public boolean isAnimal() {
		return isAnimal;
	}

	public double getNecessaryQuality() {
		return necessaryQuality;
	}
}

File: web/src/model/ItemType.ts
import { ConstantStore } from "../core/ConstantStore";

export enum ItemType {
  APPLE = "APPLE",
  BREAD = "BREAD",
  AMMO = "AMMO",
  GUN = "GUN",
  MEAT = "MEAT",
  SONIC = "SONIC",
  WAGON = "WAGON",
  WHEEL = "WHEEL",
  OX = "OX",
  TOOLS = "TOOLS",
  AXLE = "AXLE",
  HORSE = "HORSE",
  MULE = "MULE",
  STRANGEMEAT = "STRANGE_MEAT",
  MAP = "MAP",
  TRADEGOODS = "TRADEGOODS",
}

interface ItemTypeData {
  readonly name: string;
  readonly pluralName: string;
  readonly description: string;
  readonly cost: number;
  readonly weight: number;
  readonly isFood: boolean;
  readonly isPlant: boolean;
  readonly isAnimal: boolean;
  readonly isTool: boolean;
  readonly factor: number;
  readonly necessaryQuality: number;
}

const buildItemTypeData = (
  type: string,
  isFood: boolean,
  isPlant: boolean,
  isAnimal: boolean,
  isTool: boolean,
): ItemTypeData => {
  const name = ConstantStore.get("ITEMS", `${type}_NAME`);
  const pluralName = ConstantStore.get("ITEMS", `${type}_PLURAL_NAME`);
  const description = ConstantStore.get("ITEMS", `${type}_DESCRIPTION`);
  const cost = parseInt(ConstantStore.get("ITEMS", `${type}_COST`), 10);
  const weight = parseFloat(ConstantStore.get("ITEMS", `${type}_WEIGHT`));

  let factor = 0;
  if (isFood) {
    const factorStr = ConstantStore.get("ITEMS", `${type}_FOOD_FACTOR`);
    factor = factorStr ? parseInt(factorStr, 10) : 0;
  } else if (isTool) {
    const factorStr = ConstantStore.get("ITEMS", `${type}_REPAIR_FACTOR`);
    factor = factorStr ? parseInt(factorStr, 10) : 0;
  }

  const necessaryQualityStr = ConstantStore.get(
    "ITEMS",
    `${type}_NECESSARY_QUALITY`,
  );
  const necessaryQuality = necessaryQualityStr
    ? parseInt(necessaryQualityStr, 10)
    : 0;

  return {
    name,
    pluralName,
    description,
    cost,
    weight,
    isFood,
    isPlant,
    isAnimal,
    isTool,
    factor,
    necessaryQuality,
  };
};

const itemTypeDataMap: Record<ItemType, ItemTypeData> = {
  [ItemType.APPLE]: buildItemTypeData("APPLE", true, true, false, false),
  [ItemType.BREAD]: buildItemTypeData("BREAD", true, false, false, false),
  [ItemType.AMMO]: buildItemTypeData("AMMO", false, false, false, false),
  [ItemType.GUN]: buildItemTypeData("GUN", false, false, false, false),
  [ItemType.MEAT]: buildItemTypeData("MEAT", true, false, false, false),
  [ItemType.SONIC]: buildItemTypeData("SONIC", false, false, false, false),
  [ItemType.WAGON]: buildItemTypeData("WAGON", false, false, false, false),
  [ItemType.WHEEL]: buildItemTypeData("WHEEL", false, false, false, true),
  [ItemType.OX]: buildItemTypeData("OX", false, false, true, false),
  [ItemType.TOOLS]: buildItemTypeData("TOOLS", false, false, false, true),
  [ItemType.AXLE]: buildItemTypeData("AXLE", false, false, false, true),
  [ItemType.HORSE]: buildItemTypeData("HORSE", false, false, true, false),
  [ItemType.MULE]: buildItemTypeData("MULE", false, false, true, false),
  [ItemType.STRANGEMEAT]: buildItemTypeData(
    "STRANGE_MEAT",
    true,
    false,
    false,
    false,
  ),
  [ItemType.MAP]: buildItemTypeData("MAP", false, false, false, false),
  [ItemType.TRADEGOODS]: buildItemTypeData(
    "TRADEGOODS",
    false,
    false,
    false,
    false,
  ),
};

export function getItemTypeName(itemType: ItemType): string {
  return itemTypeDataMap[itemType].name;
}

export function getItemTypePluralName(itemType: ItemType): string {
  return itemTypeDataMap[itemType].pluralName;
}

export function getItemTypeDescription(itemType: ItemType): string {
  return itemTypeDataMap[itemType].description;
}

export function getItemTypeCost(itemType: ItemType): number {
  return itemTypeDataMap[itemType].cost;
}

export function getItemTypeWeight(itemType: ItemType): number {
  return itemTypeDataMap[itemType].weight;
}

export function isItemTypeFood(itemType: ItemType): boolean {
  return itemTypeDataMap[itemType].isFood;
}

export function isItemTypePlant(itemType: ItemType): boolean {
  return itemTypeDataMap[itemType].isPlant;
}

export function isItemTypeAnimal(itemType: ItemType): boolean {
  return itemTypeDataMap[itemType].isAnimal;
}

export function isItemTypeTool(itemType: ItemType): boolean {
  return itemTypeDataMap[itemType].isTool;
}

export function getItemTypeFactor(itemType: ItemType): number {
  return itemTypeDataMap[itemType].factor;
}

export function getItemTypeNecessaryQuality(itemType: ItemType): number {
  return itemTypeDataMap[itemType].necessaryQuality;
}
